### PR TITLE
feat: optimize barrel imports for RSC-incompatible packages

### LINKED ISSUE #100: shadcn fail : React.createContext is not a function
https://github.com/liuhuapiaoyuan/vinext-shadcn-bug

Image

### LINKED ISSUE #137: feat: optimize barrel imports for RSC-incompatible packages
## Problem

Barrel imports from packages like `radix-ui`, `lucide-react`, `recharts`, and others cause crashes in React Server Components (RSC) environments:

```
React.createContext is not a function
```

**Root cause:** Vite's dev server eagerly evaluates all transitive modules in a barrel. When `import { Slot } from "radix-ui"` is encountered, Vite loads all 30+ re-exported sub-packages. Some of these (e.g., `@radix-ui/react-direction`) call `React.createContext()` at module top-level, which doesn't exist in the `react-server` condition.

Next.js avoids this via `experimental.optimizePackageImports` — SWC rewrites barrel imports to direct sub-module imports at compile time.

Reported in #100.

## Solution

Add a `vinext:optimize-imports` Vite plugin that rewrites barrel imports into direct sub-module imports on the server:

```
import { Slot } from "radix-ui"
→ import * as Slot from "@radix-ui/react-slot"
```

The plugin:
- Parses barrel entry files to build an export map (which export comes from which sub-package)
- Handles all re-export patterns: `export * as X`, `export { A, B }`, `export { default as X }`, `import * as X; export { X }`
- Uses a `resolveId` hook so sub-packages resolve correctly under pnpm strict hoisting
- Only runs on server environments (RSC/SSR) — the client uses Vite's dep optimizer which handles barrels correctly
- Respects `experimental.optimizePackageImports` from `next.config` and includes sensible defaults (same list as Next.js)

## Packages optimized by default

`lucide-react`, `date-fns`, `lodash-es`, `rxjs`, `@headlessui/react`, `@heroicons/react/20/solid`, `@heroicons/react/24/solid`, `@heroicons/react/24/outline`, `@tremor/react`, `@mui/material`, `@mui/icons-material`, `recharts`, `react-bootstrap`, `antd`, `@ant-design/icons`, `ahooks`, `@tabler/icons-react`, `radix-ui`

### PR DESCRIPTION
**Pull request by gentritbiba**
## Summary

- Adds `vinext:optimize-imports` Vite plugin that rewrites barrel imports into direct sub-module imports on the server
- Prevents RSC crashes caused by eager evaluation of barrel re-exports that call `React.createContext()` at module top-level
- Parses barrel entry files to build export maps, handling all re-export patterns (`export * as X`, `export { A, B }`, `export { default as X }`, `import * as X; export { X }`)
- Uses Vite's own resolver via `resolveId` hook for pnpm strict hoisting compatibility
- Only applies on server environments (RSC/SSR) — client uses Vite's dep optimizer
- Respects `experimental.optimizePackageImports` from `next.config` with sensible defaults matching Next.js
- Lazy-initialized `Set<string>` for O(1) package lookups

Closes #100
Closes #137

## Test plan

- [x] 19 new unit tests covering plugin existence, guard clauses, all barrel export patterns, caching, aliased exports, mixed exports, and edge cases (syntax errors, unresolvable packages)
- [x] All 2030 existing tests pass
- [x] Manual verification with shadcn reproduction repo — page renders without `React.createContext` crash
- [x] Lint clean (oxlint) and type check clean (tsgo)

### CODE REVIEW

File: packages/vinext/src/index.ts [hunk 1]
```diff
@@ -193,6 +193,261 @@ function extractStaticValue(node: any): unknown {
...
+ * Used by the resolveId hook to resolve sub-packages from the barrel's context
+ * (needed for pnpm strict mode where sub-packages aren't hoisted).
+ */
+const barrelSubpkgOrigin = new Map<string, string>();
+
+/**
+ * Packages whose barrel imports are automatically optimized.
+ * Matches Next.js's built-in optimizePackageImports defaults plus radix-ui.
+ */
+const DEFAULT_OPTIMIZE_PACKAGES: string[] = [
```
Comment (zebp) [inline packages/vinext/src/index.ts]: Doesn't seem to be the exhaustive list of optimized packages
https://github.com/vercel/next.js/blob/9c31bbdaa9cf0e7674c79fe953ede9d590b5f6d7/packages/next/src/server/config.ts#L1301

**Review by zebp — COMMENTED** [review]
Need to look closer at the next implementation before I finalize the review, but so far this looks good

**Comment by gentritbiba** [comment]
Updated the `DEFAULT_OPTIMIZE_PACKAGES` list to include all packages from Next.js's defaults — added the full `effect` ecosystem (17 packages) and all `react-icons/*` sub-packages (32 packages). The list now matches [Next.js's config](https://github.com/vercel/next.js/blob/9c31bbdaa/packages/next/src/server/config.ts#L1301) exactly, plus `radix-ui` as an intentional addition.

**Comment by hyoban** [comment]
@james-elicx @southpolesteve Sorry to bother you, but could you have the bot review this again? It would help improve compatibility.

File: tests/optimize-imports.test.ts [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+  return typeof hook === "function" ? hook : hook?.handler;
+}
+
+/** Extract the vinext:optimize-imports plugin from the plugin array */
+function getOptimizeImportsPlugin(): Plugin {
+  const plugins = vinext() as Plugin[];
+  const plugin = plugins.find((p) => p.name === "vinext:optimize-imports");
+  if (!plugin) throw new Error("vinext:optimize-imports plugin not found");
+  return plugin;
+}
```
Comment (james-elicx) [inline tests/optimize-imports.test.ts]: We could probably consume the plugin directly now that it's in its own file, instead of needing the entire vinext plugin

File: packages/vinext/src/plugins/optimize-imports.ts [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+];
+
+/**
+ * Resolve a package.json exports value to a string entry path.
+ * Prefers node → import → module → default conditions, recursing into nested objects.
+ *
+ * TODO: The "react-server" condition is increasingly common — packages like `react`,
+ * `react-dom`, and `next-intl` use it to expose RSC-compatible entry points. Since this
+ * plugin targets RSC/SSR environments, "react-server" should be added to the preferred
+ * condition list (before "node") in a future pass.
```
Comment (james-elicx) [inline packages/vinext/src/plugins/optimize-imports.ts]: it would be worth supporting this before we ship the optimisation to avoid unintended side effects

File: packages/vinext/src/plugins/optimize-imports.ts [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+      // exports["./index"] or exports["./*"]). Only exports["."] is checked here,
+      // which covers the vast majority of packages in the default list. User-provided
+      // packages with non-standard export maps may need manual sub-module imports.
+      const dotExport = pkgJson.exports["."];
+      if (dotExport) {
+        const entryPath = resolveExportsValue(dotExport);
+        if (entryPath) {
+          return path.resolve(pkgDir, entryPath).split(path.sep).join("/");
+        }
+      }
```
Comment (james-elicx) [inline packages/vinext/src/plugins/optimize-imports.ts]: could we go through each of the exports and add them to a map and then pull out the right subpath based on the import the user is doing?

File: packages/vinext/src/plugins/optimize-imports.ts [hunk 5]
```diff
@@ -0,0 +1,606 @@
...
+        if (binding) {
+          exportMap.set(exported, {
+            source: binding.source,
+            isNamespace: binding.isNamespace,
+            originalName: binding.isNamespace ? undefined : binding.originalName,
+          });
+        }
+      }
+    }
+    // export * from "sub-pkg" — not resolved eagerly (left unchanged at transform time)
```
Comment (james-elicx) [inline packages/vinext/src/plugins/optimize-imports.ts]: this feels like something worth attempting to address as part of this pr if possible?

File: packages/vinext/src/plugins/optimize-imports.ts [hunk 6]
```diff
@@ -0,0 +1,606 @@
...
+  const importBindings = new Map<
+    string,
+    { source: string; isNamespace: boolean; originalName?: string }
+  >();
+
+  for (const node of ast.body as AstBodyNode[]) {
+    if (node.type === "ImportDeclaration") {
+      const source = node.source!.value as string;
+      for (const spec of node.specifiers!) {
+        if (spec.type === "ImportNamespaceSpecifier") {
```
Comment (james-elicx) [inline packages/vinext/src/plugins/optimize-imports.ts]: big nit, but some of these nested if statements could use switch statements on spec.type ;)

File: packages/vinext/src/plugins/optimize-imports.ts [hunk 7]
```diff
@@ -0,0 +1,606 @@
...
+  // Track import bindings: local name → { source, isNamespace, originalName }
+  const importBindings = new Map<
+    string,
+    { source: string; isNamespace: boolean; originalName?: string }
+  >();
+
+  for (const node of ast.body as AstBodyNode[]) {
+    if (node.type === "ImportDeclaration") {
+      const source = node.source!.value as string;
+      for (const spec of node.specifiers!) {
```
Comment (james-elicx) [inline packages/vinext/src/plugins/optimize-imports.ts]: it would be nice to remove some of the non-null assertions and type assertions in this file, either through proper type narrowing or more defensive checks inline. would be helpful for preventing unexpected crashes

File: packages/vinext/src/plugins/optimize-imports.ts [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+              // different versions; that edge case is out of scope for the default package list.)
+              barrelCaches.subpkgOrigin.set(entry.source, barrelEntry);
+            }
+          }
+
+          // Check if ALL specifiers can be resolved. If any can't, leave the import unchanged.
+          const specifiers: Array<{ local: string; imported: string }> = [];
+          let allResolved = true;
+          for (const spec of node.specifiers!) {
+            if (spec.type === "ImportSpecifier") {
```
Comment (james-elicx) [inline packages/vinext/src/plugins/optimize-imports.ts]: big nit, but switch statement potentially ;)

File: packages/vinext/src/plugins/optimize-imports.ts [hunk 9]
```diff
@@ -0,0 +1,606 @@
...
+            // that `./chunk.js` in `/node_modules/lodash-es/index.js` becomes
+            // `/node_modules/lodash-es/chunk.js` — not resolved against the
+            // importing user file (`/app/chunk.js`).
+            // Normalize to forward slashes for cross-platform safety (Windows
+            // uses backslashes in path.resolve output).
+            // TODO: barrel sources without extensions (e.g. `"./chunk"`) produce
+            // extensionless absolute paths (e.g. `/node_modules/lodash-es/chunk`).
+            // Vite's resolver handles extension resolution on these paths, so this
+            // works in practice, but a future improvement would be to resolve the
+            // extension here (or verify via the barrel AST that the file exists).
```
Comment (james-elicx) [inline packages/vinext/src/plugins/optimize-imports.ts:852]: if it's not too much work, addressing this to-do would be nice, but not a blocker

File: packages/vinext/src/plugins/optimize-imports.ts [hunk 10]
```diff
@@ -0,0 +1,606 @@
...
+
+        if (!hasChanges) return null;
+
+        return {
+          code: s.toString(),
+          map: s.generateMap({ hires: "boundary" }),
+        };
+      },
+    },
+  } as Plugin;
```
Comment (james-elicx) [inline packages/vinext/src/plugins/optimize-imports.ts]: can we remove this type assertion?

File: README.md [hunk 11]
```diff
@@ -480,12 +480,13 @@ Every `next/*` import is shimmed to a Vite-compatible implementation.
...
+| `rewrites` / `redirects` / `headers`             | ✅  | All phases, param interpolation                                                                                                                                                                                       |
+| Environment variables (`.env*`, `NEXT_PUBLIC_*`) | ✅  | Auto-loads Next.js-style dotenv files; only public vars are inlined                                                                                                                                                   |
+| `images` config                                  | 🟡  | Parsed but not used for optimization                                                                                                                                                                                  |
+| `experimental.optimizePackageImports`            | ✅  | Rewrites barrel imports to direct sub-module imports in RSC/SSR environments. A default set (lucide-react, date-fns, radix-ui, antd, MUI, and others) is always optimized. Add package names here to extend the list. |
```
Comment (james-elicx) [inline README.md]:
```suggestion
| `experimental.optimizePackageImports`            | ✅  | Rewrites barrel imports to direct sub-module imports in RSC/SSR environments. A default set (lucide-react, date-fns, radix-ui, antd, MUI, and others) are always optimized. Add package names here to extend the list. |
```

File: packages/vinext/src/config/next-config.ts [hunk 12]
```diff
@@ -501,6 +504,11 @@ export async function resolveNextConfig(
     serverActionsConfig?.bodySizeLimit as string | number | undefined,
   );
 
+  // Resolve optimizePackageImports from experimental config
+  const optimizePackageImports = Array.isArray(experimental?.optimizePackageImports)
+    ? (experimental.optimizePackageImports as string[])
```
Comment (james-elicx) [inline packages/vinext/src/config/next-config.ts]: is this type assertion needed?

File: packages/vinext/src/index.ts [hunk 13]
```diff
@@ -4511,5 +4521,6 @@ export { _postcssCache };
 export { hasMdxFiles as _hasMdxFiles };
 export { _mdxScanCache };
 export { parseStaticObjectLiteral as _parseStaticObjectLiteral };
+export { buildBarrelExportMap as _buildBarrelExportMap };
```
Comment (james-elicx) [inline packages/vinext/src/index.ts]: i dont think we need to export this from index anymore now that the plugin has its own file?
